debezium-storage/debezium-storage-jdbc/src/main/java/io/debezium/storage/jdbc/JdbcConnection.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright Debezium Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.debezium.storage.jdbc;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.SQLException;
+import java.sql.SQLRecoverableException;
+import java.time.Duration;
+
+import org.apache.kafka.connect.errors.ConnectException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import io.debezium.util.DelayStrategy;
+
+/**
+ * Establishes a new connection to JDBC
+ *
+ * @author Ilyas Ahsan
+ */
+public class JdbcConnection implements AutoCloseable {
+    private static final Logger LOGGER = LoggerFactory.getLogger(JdbcConnection.class);
+
+    private volatile Connection conn;
+    private final String jdbcUrl;
+    private final String user;
+    private final String password;
+    private final Duration waitRetryDelay;
+    private final Integer maxRetryCount;
+
+    public JdbcConnection(String jdbcUrl, String user, String password, Duration waitRetryDelay, int maxRetryCount) {
+        this.jdbcUrl = jdbcUrl;
+        this.user = user;
+        this.password = password;
+        this.waitRetryDelay = waitRetryDelay;
+        this.maxRetryCount = maxRetryCount;
+        create();
+    }
+
+    public synchronized void execute(JdbcConnectionHandler handler, boolean rollback) {
+        try {
+            handler.apply(conn);
+        }
+        catch (SQLRecoverableException e) {
+            handleRecoverableException(e);
+        }
+        catch (SQLException e) {
+            LOGGER.error("Exception while executing SQL", e);
+            throw new ConnectException("Failed to execute SQL", e);
+        }
+        finally {
+            if (rollback) {
```

Comment:
this is certainly wrong. Rollback should be done only within a catch Exception block (as you can see in my example: https://github.com/sherpa003/debezium/commit/6719ed1582dfc14fb58b6708a556f90d02ea638b)